src/AltinnCore/Common/Services/Implementation/DataAppSI.cs

```diff
@@ -33,40 +39,56 @@ public class DataAppSI : IData
         /// </summary>
         /// <param name="platformSettings">the platform settings</param>
         /// <param name="logger">the logger</param>
-        public DataAppSI(IOptions<PlatformSettings> platformSettings, ILogger<DataAppSI> logger)
+        /// <param name="httpContextAccessor">The http context accessor </param>
+        /// <param name="cookieOptions">The cookie options </param>
+        /// <param name="httpClientAccessor">The http client accessor </param>
+        public DataAppSI(
+            IOptions<PlatformSettings> platformSettings,
+            ILogger<DataAppSI> logger,
+            IHttpContextAccessor httpContextAccessor,
+            IOptions<JwtCookieOptions> cookieOptions,
+            IHttpClientAccessor httpClientAccessor)
         {
             _platformSettings = platformSettings.Value;
             _logger = logger;
+            _httpContextAccessor = httpContextAccessor;
+            _cookieOptions = cookieOptions.Value;
+            _client = httpClientAccessor.StorageClient;
         }
 
         /// <inheritdoc />
         public async Task<Instance> InsertData<T>(T dataToSerialize, Guid instanceId, Type type, string applicationOwnerId, string applicationId, int instanceOwnerId)
         {
-            string apiUrl = $"{_platformSettings.GetApiStorageEndpoint}instances/{instanceId}/data?formId={FORM_ID}&instanceOwnerId={instanceOwnerId}";
+            string apiUrl = $"instances/{instanceId}/data?formId={FORM_ID}&instanceOwnerId={instanceOwnerId}";
+            string token = JwtTokenUtil.GetTokenFromContext(_httpContextAccessor.HttpContext, _cookieOptions.Cookie.Name);
             Instance instance;
-            using (HttpClient client = new HttpClient())
+
+            _client.DefaultRequestHeaders.Accept.Clear();
+            _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/xml"));
```

Comment:
defaultrequestheaders that are common to all the clients could be moved to the httpclientaccessor you created

Reply:
dataAppSI uses application/xml, but instanceEvents and Instances uses application/json, so I'm guessing we shouldn't assign it in the accessor. Unless both can be defined in the header?

Reply:
possible to add both.